cmd/registry/cmd/annotate_test.go

```diff
@@ -0,0 +1,180 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"testing"
+
+	"github.com/apigee/registry/connection"
+	"github.com/apigee/registry/rpc"
+	"github.com/google/go-cmp/cmp"
+	"github.com/spf13/cobra"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func TestAnnotate(t *testing.T) {
+	var err error
+
+	projectId := "annotate-test"
+	projectName := "projects/" + projectId
+	apiId := "sample"
+	apiName := projectName + "/apis/" + apiId
+	versionId := "1.0.0"
+	versionName := apiName + "/versions/" + versionId
+	specId := "openapi.json"
+	specName := versionName + "/specs/" + specId
+
+	// Create a registry client.
+	ctx := context.Background()
+	registryClient, err := connection.NewClient(ctx)
+	if err != nil {
+		t.Fatalf("error creating client: %+v", err)
+	}
+	defer registryClient.Close()
+	// Clear the test project.
+	err = registryClient.DeleteProject(ctx, &rpc.DeleteProjectRequest{
+		Name: projectName,
+	})
+	if err != nil && status.Code(err) != codes.NotFound {
+		t.Fatalf("error deleting test project: %+v", err)
+	}
+	// Create the test project.
+	_, err = registryClient.CreateProject(ctx, &rpc.CreateProjectRequest{
+		ProjectId: projectId,
+		Project: &rpc.Project{
+			DisplayName: "Test",
+			Description: "A test catalog",
+		},
+	})
+	if err != nil {
+		t.Fatalf("error creating project %s", err)
+	}
+	// Create a sample api.
+	_, err = registryClient.CreateApi(ctx, &rpc.CreateApiRequest{
+		Parent: projectName,
+		ApiId:  apiId,
+		Api:    &rpc.Api{},
+	})
+	if err != nil {
+		t.Fatalf("error creating api %s", err)
+	}
+	// Create a sample version.
+	_, err = registryClient.CreateApiVersion(ctx, &rpc.CreateApiVersionRequest{
+		Parent:       apiName,
+		ApiVersionId: versionId,
+		ApiVersion:   &rpc.ApiVersion{},
+	})
+	if err != nil {
+		t.Fatalf("error creating version %s", err)
+	}
+	// Create a sample spec.
+	_, err = registryClient.CreateApiSpec(ctx, &rpc.CreateApiSpecRequest{
+		Parent:    versionName,
+		ApiSpecId: specId,
+		ApiSpec: &rpc.ApiSpec{
+			MimeType: "application/x.openapi;version=3.0.0",
+			Contents: []byte(`{"openapi": "3.0.0", "info": {"title": "test", "version": "v1"}, "paths": {}}`),
+		},
+	})
+	if err != nil {
+		t.Fatalf("error creating spec %s", err)
+	}
+
+	testCases := []struct {
+		comment  string
+		args     []string
+		expected map[string]string
+	}{
+		{"add some annotations",
+			[]string{"a=1", "b=2"},
+			map[string]string{"a": "1", "b": "2"}},
+		{"remove one annotation and overwrite the other",
+			[]string{"a=3", "b-", "--overwrite"},
+			map[string]string{"a": "3"}},
+		{"changing an annotation without --overwrite should be ignored",
+			[]string{"a=4"},
+			map[string]string{"a": "3"}},
+	}
+	// test annotations for APIs.
+	for _, tc := range testCases {
+		cmd := &cobra.Command{}
+		cmd.SetArgs(append([]string{"annotate", apiName}, tc.args...))
+		annotateCmd(cmd)
```

Comment:
@seaneganx see these three lines (114-116) for an alternate way to construct command handlers. I don't think this would be the final form - eventually I think it would be good to have a rootCmd() function that explicitly builds the full command tree instead of relying on init methods. (The init methods make commands easy to add locally but they don't allow us to construct multiple identical rootCmd() trees, which would help in testing and possibly in the implementation of the controller (suppose the controller used worker goroutines to run multiple commands in parallel).

A future revision might replace these three lines with 
```
cmd := rootCmd(append([]string{"annotate", apiName}, tc.args...)
```
where rootCmd automatically includes the annotateCmd (and all others).

Reply:
Agreed, I actually started poking around with that idea yesterday. I didn't think about accepting arguments in the constructors. It might be convenient, but we still have cmd.SetArgs to achieve the same thing more clearly?

The pattern I came up with yesterday took advantage of packages more than our existing solution, so creating the annotate command was a function called `annotate.Command()` rather than `cmd.annotateCmd(parent *cobra.Command)`. With that approach the child is isolated until a parent command attaches it, allowing us to avoid the mandatory "parent" attachment if we want to test in isolation.

I can put up a draft PR if you're interested.